File: excel/CharacterData.java
package excel;

import java.io.IOException;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;

import org.apache.poi.xssf.usermodel.XSSFCellStyle;
import org.apache.poi.xssf.usermodel.XSSFSheet;
import org.apache.poi.xssf.usermodel.XSSFWorkbook;

import models.Character;
import models.lists.Characters;
import models.lists.Lists;
import models.lists.Spells;

public class CharacterData
{
    public static void makeChangeList()
    {
	try (XSSFWorkbook workbook = new XSSFWorkbook(); OutputStream fileOut = Excel.createOutputStream("Character.xlsx"))
	{

	    XSSFSheet sheet = workbook.createSheet("Character Data");

	    XSSFCellStyle defaultStyle = Excel.setDefaultStyle(workbook);
	    XSSFCellStyle boldStyle = Excel.setBoldStyle(workbook);

	    List<RowData> rowData = new ArrayList<RowData>();

	    Characters characters = new Characters();
	    characters.addUsefulModels(Lists.getCharacters(), false);

	    for (int i = 0; i < Characters.characterListSize; i++)
	    {

		Character c = characters.getModels().get(i);
		Character Dc = characters.getDModels().get(i);

		rowData.add(new RowData(boldStyle, new String[]
		{
			c.getName()
		}));

		rowData.add(new RowData(defaultStyle, new String[]
		{
			"Stats", "HP", "MP", "Power", "Guard", "Magic", "Speed"
		}));
		rowData.add(new RowData(defaultStyle, new String[]
		{
			"Initial", Excel.CompareValues(Dc.getHpStart(), c.getHpStart()), Excel.CompareValues(Dc.getMpStart(), c.getMpStart()),
			Excel.CompareValues(Dc.getPowerStart(), c.getPowerStart()), Excel.CompareValues(Dc.getGuardStart(), c.getGuardStart()),
			Excel.CompareValues(Dc.getMagicStart(), c.getMagicStart()), Excel.CompareValues(Dc.getSpeedStart(), c.getSpeedStart())
		}));

		rowData.add(new RowData(defaultStyle, new String[]
		{
			"Level Up", Excel.CompareValues(Dc.getHpGrowth(), c.getHpGrowth()), Excel.CompareValues(Dc.getMpGrowth(), c.getMpGrowth()),
			Excel.CompareValues(Dc.getPowerGrowth(), c.getPowerGrowth()), Excel.CompareValues(Dc.getGuardGrowth(), c.getGuardGrowth()),
			Excel.CompareValues(Dc.getMagicGrowth(), c.getMagicGrowth()), Excel.CompareValues(Dc.getSpeedGrowth(), c.getSpeedGrowth()),
		}));

		rowData.add(new RowData(defaultStyle, new String[]
		{
			""
		}));

		List<String> spells = new ArrayList<String>();
		List<String> levels = new ArrayList<String>();

		List<String> spells2 = new ArrayList<String>();
		List<String> levels2 = new ArrayList<String>();

		spells.add("Spells");
		levels.add("Levels");

		spells2.add("Spells");
		levels2.add("Levels");

		Spells spellsList = Lists.getSpells();

		List<String> characterSpells = spellsList.getNames(c.getSpells());
		List<String> DCharacterSpells = spellsList.getDNames(Dc.getSpells());

		for (int j = 0; j < characterSpells.size(); j++)
		{
		    characterSpells.set(j, Excel.modifyCase(characterSpells.get(j)));
		}

		for (int j = 0; j < DCharacterSpells.size(); j++)
		{
		    DCharacterSpells.set(j, Excel.modifyCase(DCharacterSpells.get(j)));
		}

		for (int k = 0; k < characterSpells.size(); k++)
		{
		    if (k < DCharacterSpells.size())
		    {
			spells.add(Excel.CompareValues(DCharacterSpells.get(k), characterSpells.get(k)));
			levels.add(Excel.CompareValues(Dc.getSpellLevels().get(k), c.getSpellLevels().get(k)));
		    }
		    else
		    {
			spells.add(Excel.CompareValues("", characterSpells.get(k)));
			levels.add(Excel.CompareValues("", Integer.toString(c.getSpellLevels().get(k))));
		    }
		}

		while (spells.size() > 9)
		{
		    spells2.add(spells.remove(9));
		}
		while (levels.size() > 9)
		{
		    levels2.add(levels.remove(9));
		}

		rowData.add(new RowData(defaultStyle, spells.toArray(new String[spells.size()])));
		rowData.add(new RowData(defaultStyle, levels.toArray(new String[levels.size()])));

		if (spells2.size() > 1)
		{
		    rowData.add(new RowData(defaultStyle, new String[]
		    {
			    ""
		    }));
		    rowData.add(new RowData(defaultStyle, spells2.toArray(new String[spells2.size()])));
		    rowData.add(new RowData(defaultStyle, levels2.toArray(new String[levels2.size()])));

		}

		rowData.add(new RowData(defaultStyle, new String[]
		{
			""
		}));
	    }
	    Excel.writeDataToSheet(sheet, rowData);

	    workbook.write(fileOut);

	}
	catch (IOException e)
	{
	    e.printStackTrace();
	}
    }

}

File: src/application/excel/CharacterData.kt
package application.excel

import application.models.lists.Characters
import application.models.lists.Lists
import org.apache.poi.xssf.usermodel.XSSFSheet

class CharacterData(fileName: String) : Excel(fileName)
{
	override fun makeChangeList()
	{
		val characters = Characters()
		characters.addUsefulModels(Lists.characters, false)
		characters.chronologicalIndexSort()
		fillStatsSheet(characters)
		fillSpellsSheet(characters)

	}
	private fun fillStatsSheet(characters: Characters) {
		val statsSheet = workbook.createSheet("Stats")
		val sheetData = SheetData()
		sheetData.addRow(RowData(bigBoldStyle, mutableListOf("", "HP", "MP", "Power", "Guard", "Magic", "Speed")))
		sheetData.addRow(RowData(bigStyle, mutableListOf("Starting Stats")))
		characters.models.forEachIndexed { index, c ->
			val Dc = characters.dModels[index]
			sheetData.addRow(RowData(defaultStyle, mutableListOf(
				c.name,
				CompareValues(Dc.hpStart,c.hpStart),
				CompareValues(Dc.mpStart,c.mpStart),
				CompareValues(Dc.powerStart,c.powerStart),
				CompareValues(Dc.guardStart,c.guardStart),
				CompareValues(Dc.magicStart, c.magicStart),
				CompareValues(Dc.speedStart, c.speedStart)
			)))
		}
		sheetData.addRow(RowData(defaultStyle, mutableListOf("")))
		sheetData.addRow(RowData(bigBoldStyle, mutableListOf("Growth Rates")))
		characters.models.forEachIndexed { index, c ->
			val Dc = characters.dModels[index]
			sheetData.addRow(RowData(defaultStyle, mutableListOf(
				c.name,
				CompareValues(Dc.hpGrowth,c.hpGrowth),
				CompareValues(Dc.mpGrowth,c.mpGrowth),
				CompareValues(Dc.powerGrowth,c.powerGrowth),
				CompareValues(Dc.guardGrowth,c.guardGrowth),
				CompareValues(Dc.magicGrowth, c.magicGrowth),
				CompareValues(Dc.speedGrowth, c.speedGrowth)
			)))
		}
		statsSheet.createFreezePane(0,1,0,1)
		sheetData.setColumnStyle(bigBoldStyle,0)
		writeDataToSheet(statsSheet, sheetData)
	}


	private fun fillSpellsSheet(characters: Characters) {
		val spellsSheet = workbook.createSheet("Spells")
		val sheetData = SheetData()
		val header = mutableListOf<String>()
		header.add("#")
		characters.models.forEach{

			header.add(it.name)
			header.add("Level")
			header.add(" ")
		}

		sheetData.addRow(RowData(boldStyle, header))
		val spells = Lists.spells
		for(i in 0 until 16){
			val row = mutableListOf<String>()
			row.add((i+1).toString())

			characters.models.forEachIndexed{ index, c ->
				val Dc = characters.dModels[index]


				row.add(CompareValues(spells.getDName(Dc.spells[i]), spells.getName(c.spells[i])))
				row.add(CompareValues(Dc.spellLevels[i], c.spellLevels[i]))
				row.add("")
			}
			sheetData.addRow(RowData(defaultStyle, row))
		}
		spellsSheet.createFreezePane(1,1,1,1)
		sheetData.setColumnStyle(bigBoldStyle,0)
		writeDataToSheet(spellsSheet, sheetData)
	}
	/*	for (i in 0..Characters.characterListSize - 1)
		{

			val c = characters.models[i]
			val Dc = characters.dModels[i]

			sheetData.addRow(RowData(boldStyle, mutableListOf(c.name)))

			sheetData.addRow(RowData(defaultStyle, mutableListOf("Initial", CompareValues(Dc.hpStart, c.hpStart), CompareValues(Dc.mpStart, c.mpStart), CompareValues(Dc.powerStart, c.powerStart), CompareValues(Dc.guardStart, c.guardStart), CompareValues(Dc.magicStart, c.magicStart), CompareValues(Dc.speedStart, c.speedStart))))
			sheetData.addRow(RowData(defaultStyle, mutableListOf("Level Up", CompareValues(Dc.hpGrowth, c.hpGrowth), CompareValues(Dc.mpGrowth, c.mpGrowth), CompareValues(Dc.powerGrowth, c.powerGrowth), CompareValues(Dc.guardGrowth, c.guardGrowth), CompareValues(Dc.magicGrowth, c.magicGrowth), CompareValues(Dc.speedGrowth, c.speedGrowth))))
			sheetData.addRow(RowData(defaultStyle, mutableListOf("")))

			val spells = mutableListOf<String>()
			val levels = mutableListOf<String>()

			val spells2 = mutableListOf<String>()
			val levels2 = mutableListOf<String>()

			spells.add("Spells")
			levels.add("Levels")

			spells2.add("Spells")
			levels2.add("Levels")

			val spellsList = Lists.spells

			val characterSpells = spellsList.getNames(c.spells)
			val DCharacterSpells = spellsList.getDNames(Dc.spells)

			characterSpells.indices.forEach { characterSpells[it] = modifyCase(characterSpells[it]) }
			DCharacterSpells.indices.forEach { DCharacterSpells[it] = modifyCase(DCharacterSpells[it]) }

			characterSpells.indices.forEach { k ->
				if (k < DCharacterSpells.size)
				{
					spells.add(CompareValues(DCharacterSpells[k], characterSpells[k]))
					levels.add(CompareValues(Dc.spellLevels[k], c.spellLevels[k]))
				}
				else
				{
					spells.add(CompareValues("", characterSpells[k]))
					levels.add(CompareValues("", Integer.toString(c.spellLevels[k])))
				}
			}

			while (spells.size > 9)
			{
				spells2.add(spells.removeAt(9))
			}
			while (levels.size > 9)
			{
				levels2.add(levels.removeAt(9))
			}

			sheetData.addRow(RowData(defaultStyle, spells))
			sheetData.addRow(RowData(defaultStyle, levels))

			if (spells2.size > 1)
			{
				sheetData.addRow(RowData(defaultStyle, mutableListOf("")))
				sheetData.addRow(RowData(defaultStyle, spells2))
				sheetData.addRow(RowData(defaultStyle, levels2))
			}

			sheetData.addRow(RowData(defaultStyle, mutableListOf("")))
		}
		writeDataToSheet(statsSheet, sheetData)*/

}